packages/coinjoin/src/client/round/selectRound.ts

```diff
@@ -91,6 +93,7 @@ export const getAccountCandidates = (
                 ) {
                     account.skipRoundCounter = 0;
                     log(`Random skip candidate ~~${accountKey}~~`);
+                    hasSkippedRounds = true;
```

Comment:
woudnt be better to use `CoinjoinRoundOptions.setSessionPhase` here instead of changing whole function?

Reply:
But `SessionPhase.RetryingCoinSelection` would be emitted anyways in that case since it simply checks for `accountCandidates.length < 1` :( And I was trying another solution without changing the function but its result depends on `Math.random()`, therefore I need to know the exact result from within `getAccountCandidates`. And actually, a round being skipped does not mean that `accountCandidates.length < 1` will be true, does it? And it's not that beneficial to know a round is randomly skipped when the whole process goes on without an error.

Reply:
current solution doesn't distinguish between the reason and the accounts,
so if one account was not picked because of random skip, second account was not picked because it have no utxos and third account was not picked because its already registered then they all will receive SkippingRound, moreover you need to introduce more logic around it (like getIsAlreadyRegistered) 


My suggestion for changes in `getAccountCandidates` (pseudo code):
```
const skippedAccounts: { key: string; reason: SessionPhase }[] = [];
const candidates = accounts.flatMap(account => {
    ...[logic]...
    if (utxos.length === 0) {
        skippedAccounts.push({ key: account.accountKey, reason: SessionPhase.AccountMissingUtxos });
    }
    if (skipping) {
        skippedAccounts.push({ key: account.accountKey, reason: SessionPhase.SkippingRound });
    }
}

if (skippedAccounts.length > 0) {
     setSessionPhase({
                phase: SessionPhase.SkippingRound, // TODO: reason
                accountKeys: skippedAccounts.map(a => a.accountKey),
      });
}

return candidates;

```

Reply:
I see several issues with this: `appendSessionPhase` doesn't append duplicates so this `forEach` leads to 1 event in Suite _(3s)_  and if somehow there are 3 accounts with `[AccountMissingUtxos, SkippingRound, AccountMissingUtxos]` it's gonna take 9s just to reflect 1 `account selection failed` state. And on top of that `RetryingCoinSelection` might still be dispatched, so 12s, right? If `selectRound` fails on the `getAccountCandidates` step, I think only one event should be dispatched. Since users know that some rounds are skipped _(it is mentioned on the setup screen)_, the `Round skipped` is more relevant than `Missing UTXOs`, which simply fall into the `RetryingCoinSelection` category. If indeed several accounts are not selected for different reasons I would not try to communicate all of those reasons, as it's not that beneficial for the users and would take a lot of time compared to other phases. Finally, in other situations, e.g. `SignatureFailed` there might be different reasons for an error too? I believe we do not need to be extremely detailed with the phases, @hynek-jina even mentioned that to me.

Reply:
>(...) leads to 1 event in Suite (3s) and if somehow there are 3 accounts (...)

this is why i've put comment "or you can group account keys by reason and then call setSessionPhase"

> If selectRound fails on the getAccountCandidates step, I think only one event should be dispatched.

but which one in the case i've just described? becuase in current state all 3 accounts will receive "SkippingRound" which is not true

> the Round skipped is more relevant than Missing UTXOs, which simply fall into the RetryingCoinSelection category.

we can use AccountMissingUtxos to display message when you dont have confirmed utxos (currently its misleading and people [complains ](https://satoshilabs.slack.com/archives/C048XFAKL10/p1671023775628639?thread_ts=1671021930.041649&cid=C048XFAKL10)that they are not picked to round and they dont know why)

Reply:
btw why event takes 3 sec to get to suite?